File: Power Query/400-499/400/PQ_400.R
library(tidyverse)
library(readxl)

path <- "400-499/400/PQ_Challenge_400.xlsx"
input <- read_excel(path, range = "A1:A12")
test <- read_excel(path, range = "C1:F14")


priority <- c(Blocked = 1, Review = 2, Open = 3, Closed = 4)

result <- input %>%
  separate(
    Event,
    into = c("Entity", "StartDate", "EndDate", "Status"),
    sep = ","
  ) %>%
  mutate(
    across(c(StartDate, EndDate), dmy),
    StatusRank = priority[Status],
    Date = map2(StartDate, EndDate, seq, by = "day")
  ) %>%
  unnest(Date) %>%
  slice_min(StatusRank, n = 1, with_ties = FALSE, by = c(Entity, Date)) %>%
  arrange(Entity, Date) %>%
  mutate(
    grp = cumsum(Status != lag(Status, default = first(Status))),
    .by = Entity
  ) %>%
  summarise(
    StartDate = as.POSIXct(min(Date)),
    EndDate = as.POSIXct(max(Date)),
    FinalStatus = first(Status),
    .by = c(Entity, grp)
  ) %>%
  select(Entity, StartDate, EndDate, FinalStatus)

all.equal(result, test, check.attributes = FALSE)

File: Power Query/400-499/400/PQ_400.py
import pandas as pd

path = "400-499/400/PQ_Challenge_400.xlsx"
input = pd.read_excel(path, usecols="A", nrows=12)
test = pd.read_excel(path, usecols="C:F", nrows=14)

priority = {"Blocked": 1, "Review": 2, "Open": 3, "Closed": 4}

parts = input["Event"].str.split(",", expand=True)
result = (
    input.assign(
        Entity=parts[0].str.strip(),
        StartDate=pd.to_datetime(parts[1].str.strip(), dayfirst=True, errors="coerce"),
        EndDate=pd.to_datetime(parts[2].str.strip(), dayfirst=True, errors="coerce"),
        Status=parts[3].str.strip(),
    )
    .drop(columns="Event")
    .dropna(subset=["StartDate", "EndDate"])
    .assign(
        StatusRank=lambda d: d["Status"].map(priority),
        Date=lambda d: d.apply(
            lambda r: pd.date_range(r["StartDate"], r["EndDate"], freq="D"), axis=1
        ),
    )
    .explode("Date")
    .sort_values(["StatusRank", "Entity", "Date"])
    .drop_duplicates(["Entity", "Date"])
    .sort_values(["Entity", "Date"])
    .assign(
        grp=lambda d: d.groupby("Entity")["Status"].transform(
            lambda s: s.ne(s.shift()).cumsum()
        )
    )
    .groupby(["Entity", "grp"], sort=False, as_index=False)
    .agg(
        StartDate=("Date", "min"),
        EndDate=("Date", "max"),
        FinalStatus=("Status", "first"),
    )
    .drop(columns="grp")
)

print(result.equals(test))
